src/Bitcoin/Entry.cpp

```diff
@@ -104,6 +106,55 @@ string Entry::deriveAddress(TWCoinType coin, TWDerivation derivation, const Publ
     }
 }
 
+Data Entry::addressToData(TWCoinType coin, const std::string& address) const {
+    switch (coin) {
+        case TWCoinTypeBitcoin:
+        case TWCoinTypeDigiByte:
+        case TWCoinTypeGroestlcoin:
+        case TWCoinTypeLitecoin:
+        case TWCoinTypeViacoin:
+            {
+                const auto decoded = SegwitAddress::decode(address);
+                if (!std::get<2>(decoded)) {
+                    return Data();
+                }
+                return std::get<0>(decoded).witnessProgram;
+            }
+
+        case TWCoinTypeBitcoinCash:
+            {
+                const auto addr = BitcoinCashAddress(address);
+                Data data(Address::size);
+                size_t outlen = 0;
+                cash_data_to_addr(data.data(), &outlen, addr.bytes.data(), 34);
+                return {data.begin() + 1, data.end()};
+            }
+
+        case TWCoinTypeECash:
+            {
+                const auto addr = ECashAddress(address);
+                Data data(Address::size);
+                size_t outlen = 0;
+                cash_data_to_addr(data.data(), &outlen, addr.bytes.data(), 34);
+                return {data.begin() + 1, data.end()};
+            }
```

Comment:
The code could be simplified with a generic function:

```cpp
template<typename Address>
Data cashAddressToData(Address&& addr) {
    Data data(Address::size);
    size_t outlen = 0;
    cash_data_to_addr(data.data(), &outlen, addr.bytes.data(), 34);
    return {data.begin() + 1, data.end()};
}

return cashAddressToData(BitcoinCashAddress(address));
return cashAddressToData(ECashAddress(address));
```

Could also use `auto&&` as parameter instead of template if you prefer

Reply:
In this case, it turns out, part of the logic could be put in the common base class (`CashAddress`).
The proposed template would have been a good solution, the drawback in my eye is that it has to be defined in the `.h` file (being a template), which means the needed `trezor-crypto` header has to be include in the `.h` file, 'polluting' the compilation of all the source files including it (indirectly many do), slowing down compilation.

Reply:
https://github.com/trustwallet/wallet-core/pull/2300/commits/01d084d7687d6e1c7f216a2577b72342d23405b4